Ackermann: Refactor

chfriedrich98 opened the pull request:
<!--

Thank you for your contribution!

Get early feedback through
- Dronecode Discord: https://discord.gg/dronecode
- PX4 Discuss: http://discuss.px4.io/
- opening a draft pr and sharing the link

-->

### Solved Problem
This PR starts a restructure of the rover modules to the following architecture:
![rover_code_architecture](https://github.com/user-attachments/assets/aaac9601-c88a-41d5-9c7a-92a73bc600d5)

This PR focuses on restructuring the ackermann rover module (differential and mecanum will be updated in subsequent PRs).

It also moves  functions and parameters that are shared by the different rover modules to a common rover library (`rover_control`) to reduce redundancy. This library and parameters will also be used in follow up PRs for the differential and mecanum rover modules.

In order to homogenize the 3 rover modules, the rate control for the ackermann module is transformed from lateral acceleration to yaw rate control (to get in line with differential/mecanum). This does not lead to a functional difference as lateral acceleration ($$a_y$$) and yaw rate ($$\dot{\psi}$$) are direclty linked with this equation: $$a_y = V \cdot \dot{\psi}$$ with $$V: $$ Forward speed.

This PR also introduces offboard support for ackermann rovers through the `trajectorySetpoint.msg`. (Works on [#23663](https://github.com/PX4/PX4-Autopilot/issues/23663))
The Ackermann module supports **position** and **velocity** setpoints. 

### Test coverage
- SITL Tested
- Hardware tests:
  - Manual Mode: https://review.px4.io/plot_app?log=5f4d4f03-e0e7-44bb-bf34-3f43d873c481
  - Acro Mode: https://review.px4.io/plot_app?log=cd97eab4-4e4a-43e5-ad8e-afc071ed721f
![image](https://github.com/user-attachments/assets/d791eb8a-99e9-49bb-8cb3-a5b76c0585f4)
  - Stab Mode: https://review.px4.io/plot_app?log=77580a37-34c6-4eaf-8744-276283f0587e
![image](https://github.com/user-attachments/assets/620eb913-91e3-4262-9cef-466729a879ad)
  - Position Mode: https://review.px4.io/plot_app?log=dea91bbf-b271-4e04-93c3-337a8aefa3c3
![image](https://github.com/user-attachments/assets/3c9ec35e-76cd-458d-8da6-442cf517a481)
  - Auto Mode: https://review.px4.io/plot_app?log=221db1da-f0b9-4910-a5d5-04b951664a83
![image](https://github.com/user-attachments/assets/4258e86b-3281-424f-aa84-7d6dd3cdf97e)
- Offbaord Tests
  - Position (SITL): https://review.px4.io/plot_app?log=93e0e5c3-1dbc-4cd3-80b8-97f6bc37e61e
![image](https://github.com/user-attachments/assets/5bcac31e-7e47-4f29-8ce9-c8c4baf7ee40)
  - Velocity (SITL): https://review.px4.io/plot_app?log=93e0e5c3-1dbc-4cd3-80b8-97f6bc37e61e
![image](https://github.com/user-attachments/assets/862c0b8f-eb90-4caa-936e-378512bad283)

sfuhrer commented on src/modules/rover_ackermann/AckermannPosControl/AckermannPosControl.hpp:
```diff
@@ -0,0 +1,247 @@
+	float _vehicle_yaw{0.f};
+	float _max_yaw_rate{0.f};
+	float _forward_speed_setpoint{0.f};
+	float _min_speed{0.f};
```
Here you could either by renaming or with a comment make it more clear where this constraint is coming from

chfriedrich98 replied:
Added comment in [454995a](https://github.com/PX4/PX4-Autopilot/pull/24285/commits/454995a3a15f362273c6bc3da6fc9edb9e088165)

sfuhrer commented on src/modules/rover_ackermann/RoverAckermann.cpp:
```diff
@@ -53,174 +53,113 @@ bool RoverAckermann::init()
 
-	if (!_armed) {
-		_ackermann_control.resetControllers();
+	bool manual_mode_enabled = _vehicle_control_mode.flag_control_manual_enabled
```
```suggestion
	bool full_manual_mode_enabled = _vehicle_control_mode.flag_control_manual_enabled
```
To distinguish it from the other "manual" modes that still contain some controller support.

chfriedrich98 replied:
Suggestion applied in [454995a](https://github.com/PX4/PX4-Autopilot/pull/24285/commits/454995a3a15f362273c6bc3da6fc9edb9e088165)

sfuhrer commented on src/modules/rover_ackermann/AckermannPosControl/AckermannPosControl.cpp:
```diff
@@ -0,0 +1,390 @@
+	// Publish updated acceptance radius
+	position_controller_status_s pos_ctrl_status{};
+	pos_ctrl_status.acceptance_radius = acceptance_radius;
+	pos_ctrl_status.timestamp = hrt_absolute_time();
```
```suggestion
	pos_ctrl_status.timestamp = _timestamp;
```
To be consistent with the other publications

chfriedrich98 replied:
Applied in [454995a](https://github.com/PX4/PX4-Autopilot/pull/24285/commits/454995a3a15f362273c6bc3da6fc9edb9e088165)

chfriedrich98 commented on src/modules/rover_ackermann/AckermannRateControl/AckermannRateControl.cpp:
```diff
@@ -0,0 +1,180 @@
+	ModuleParams::updateParams();
+	_max_yaw_rate = _param_ro_max_yaw_rate.get() * M_DEG_TO_RAD_F;
+	_pid_yaw_rate.setGains(_param_ro_yaw_rate_p.get(), _param_ro_yaw_rate_i.get(), 0.f);
+	_pid_yaw_rate.setIntegralLimit(1.f);
```
The [default is zero](https://github.com/PX4/PX4-Autopilot/blob/main/src/lib/pid/PID.hpp#L59-L64)

chfriedrich98 commented on src/modules/rover_ackermann/AckermannAttControl/AckermannAttControl.cpp:
```diff
@@ -0,0 +1,194 @@
+
+void AckermannAttControl::generateAttitudeSetpoint()
+{
+	bool stab_mode_enabled = _vehicle_control_mode.flag_control_manual_enabled
```
Scanned all files and added the missing const prefixes here: [454995a](https://github.com/PX4/PX4-Autopilot/pull/24285/commits/454995a3a15f362273c6bc3da6fc9edb9e088165)

sfuhrer replied:
Mhh cannot find this commit anymore - is it still applied? And my comment here for a `const` was not valid?

chfriedrich98 replied:
Yes, that should be const. Fixed in [e226138](https://github.com/PX4/PX4-Autopilot/pull/24285/commits/e226138032f972288f8e336b7bbabbb480112ee2)

chfriedrich98 commented:
@sfuhrer Thanks for the review!

> as handle both position and velocity in the "Position" controller - how about renaming it to PositionVelocityControl?

Implemented that in [5cdc876](https://github.com/PX4/PX4-Autopilot/pull/24285/commits/5cdc876fa823bfe52590ac9b9b6d9518f03d6365), also renamed the parameter group `Rover Position Control` to `Rover Velocity Control` and the uORB message `RoverPositionStatus` to `RoverVelocityStatus` because they both actually only deal with velocity.

> in the diagram you show that also attitude and rate controllers subscribe to TrajectorySetpoint.msg - is this only for Differential/Mecanum?

Yes, while the refactor does technically enable setting yaw/yaw rate setpoints through the TrajectorySetpoint also for ackermann it gets messy because the yaw/yaw rate is not decoupled from the velocity (or speed) as it is in differential/mecanum rovers. 
That means the user would have to provide a speed setpoint of some sort if they want to do yaw/yaw rate control with an ackermann rover, but this is not an option through the `trajectorySetpoint.msg`.
This could be done by "hijacking" the `rover_throttle_setpoint.msg` or directly `actuator_motors.msg` but this is where it gets messy.

chfriedrich98 commented on src/lib/rover_control/module.yaml:
```diff
@@ -0,0 +1,208 @@
+          max: 100
+          increment: 0.01
+          decimal: 2
+          default: -1
```
Changed the default/min value to 0 in [72f65d2](https://github.com/PX4/PX4-Autopilot/pull/24285/commits/72f65d2aa450a58fe33a49e3eee6c95d41d5a9a9) because this parameter can't be "disabled" but rather **must** be set for the module to work.

chfriedrich98 commented on src/lib/rover_control/module.yaml:
```diff
@@ -0,0 +1,208 @@
+
+        RO_MAX_DECEL:
+          description:
+            short: Maximum deceleration for the rover
```
Improved param descriptions in [72f65d2](https://github.com/PX4/PX4-Autopilot/pull/24285/commits/72f65d2aa450a58fe33a49e3eee6c95d41d5a9a9).

chfriedrich98 commented on msg/RoverPositionStatus.msg:
```diff
@@ -0,0 +1,12 @@
+float32 measured_lateral_speed          # [m/s] Measured speed in body y direction. Positiv: right, Negativ: left
+float32 lateral_speed_setpoint          # [m/s] Speed setpoint in body y direction. Positiv: right, Negativ: left (Only for mecanum rovers)
+float32 adjusted_lateral_speed_setpoint # [m/s] Post slew rate speed setpoint in body y direction. Positiv: right, Negativ: left (Only for mecanum rovers)
+float32 pid_forward_throttle_integral   # Integral of the PID for the closed loop forward speed controller
```
Homogenized the wording to `speed_body_x` and `speed_body_y` throughout the module to clarify (it is now also inline with the `RoverThrottleSetpoint.msg`) in [82fdba0](https://github.com/PX4/PX4-Autopilot/pull/24285/commits/82fdba0134d41b429bf321d0c824bf4b7fd08ce7)

chfriedrich98 commented on src/lib/rover_control/RoverControl.cpp:
```diff
@@ -0,0 +1,208 @@
+		return NAN;
+	}
+
+	// Apply acceleration and deceleration limit
```
I would prefer keeping this in one function because I can reuse it exactly like this in the other rover modules. 
Seperating it  would just end up requiring to call both functions every time.
While the naming makes sense to me based on the code architecture (transforming a velocity setpoint to a throttle setpoint which is the interface to the next layer) I could rename it to something like `speedControl` ?

sfuhrer replied:
Fine for me with the renamed function name :+1:

chfriedrich98 replied:
Changed in https://github.com/PX4/PX4-Autopilot/pull/24285/commits/163e3fb45c92af9291d422724a5ca91947c42e04

chfriedrich98 commented:
rebased on main

sfuhrer commented on msg/RoverAttitudeStatus.msg:
```diff
@@ -0,0 +1,7 @@
+
+float32 measured_yaw          # [rad/s] Measured yaw rate
+float32 adjusted_yaw_setpoint # [rad/s] Yaw setpoint that is being tracked (Applied slew rates)
+float32 pid_yaw_integral      # Integral of the PID for the closed loop yaw controller
```
Still there? I thought you meant to remove the attitude I gain.

chfriedrich98 replied:
I removed the parameter but forgot to remove the logging. Changed in [e226138](https://github.com/PX4/PX4-Autopilot/pull/24285/commits/e226138032f972288f8e336b7bbabbb480112ee2)

chfriedrich98 commented:
@Pedro-Roque Just FYI i properly disabled the rover modules on the spacecraft board [here](https://github.com/PX4/PX4-Autopilot/pull/24285/files#diff-94ed692b00cbc0164071e7756b3419d077a7159d5fecbba3bd18c06ca0c6fa7aR14-R16).

sfuhrer commented on src/modules/rover_ackermann/module.yaml:
```diff
@@ -6,26 +6,37 @@ parameters:
         increment: 0.001
         decimal: 3
-        default: 0.5
+        default: 0
```
Doesn't have to be in this PR, but it would improve the UX if you'd publish a warning if one of the params isn't configured, instead of failing silently.
As inspiration, [this is what we have for FW.](https://github.com/PX4/PX4-Autopilot/blob/e7e76e2e2124239319768e1f3e4e7b8942797bcf/src/lib/fw_performance_model/PerformanceModel.cpp#L164)

sfuhrer commented on src/modules/rover_ackermann/module.yaml:
```diff
@@ -6,26 +6,37 @@ parameters:
+      RA_MAX_STR_RATE:
+        description:
+          short: Maximum steering rate
+          long: Set to -1 to disable.
```
What's the outcome if this is disabled?

chfriedrich98 replied:
Just FYI i renamed this in [e226138](https://github.com/PX4/PX4-Autopilot/pull/24285/commits/e226138032f972288f8e336b7bbabbb480112ee2) because i forgot to adapt it to the new naming convention i use:
I differentiate between parameters of the physical limits of the system (i.e. steering angle, speed at max throttle) and the user defined limits (i.e. maximum acceleration, speed, steering rate) (while these can be set to the physical limit of the system, in general they can be any user choosen value to protect the rover from rolling over, wheeling etc.).
To differentiate i use the `MAX` keyword for physical limits (i.e. `RA_MAX_STR_ANG`) and the `LIM` keyword for user defined limits (i.e. `RO_ACC_LIM` or `RA_STR_RATE_LIM`).
The `LIM` parameters are optional (but configuring them will greatly improve the behaviour).

Disabling the steering rate limit simply means there is no slew rate applied between the steering angle setpoint and the servo motor command. (The slew rate can be used to protect the servo from i.e. too aggresive or high frequency changes).

sfuhrer replied:
>MAX keyword for physical limits (i.e. RA_MAX_STR_ANG) and the LIM keyword for user defined 

Nice, makes sense! That makes the parameter name by itself already more understandable.

sfuhrer commented on src/modules/rover_ackermann/AckermannRateControl/AckermannRateControl.cpp:
```diff
@@ -71,13 +71,13 @@ void AckermannRateControl::updateRateControl()
 	}
 
-	if (_vehicle_control_mode.flag_control_rates_enabled  && _vehicle_control_mode.flag_armed) {
+	if (_vehicle_control_mode.flag_control_rates_enabled  && _vehicle_control_mode.flag_armed && runSanityChecks()) {
```
How about just checking `_sanity_check` here and calling `runSanityChecks` only on param update?

chfriedrich98 replied:
I agree that it is suboptimal to run this check on every update rather then on paramUpdate, but the check should also only be performed if the corresponding flag of the mode is active (i.e. only do the check for posVelControl if it is actually active). And there is the scenario where the mode is switched from i.e. acro to position without changing parameters which will then not perform the checks for posVel parameters.

sfuhrer replied:
okay makes sense

sfuhrer commented on src/modules/rover_ackermann/AckermannRateControl/AckermannRateControl.hpp:
```diff
@@ -112,6 +119,7 @@ class AckermannRateControl : public ModuleParams
 	float _vehicle_yaw_rate{0.f};
 	hrt_abstime _timestamp{0};
 	float _dt{0.f}; // Time since last update [s].
+	bool _sanity_check{true};
```
I like when bools are very clearly visible as such - e.g. _parameter_sanity_checks_passed

Pedro-Roque commented:
> @Pedro-Roque Just FYI i properly disabled the rover modules on the spacecraft board [here](https://github.com/PX4/PX4-Autopilot/pull/24285/files#diff-94ed692b00cbc0164071e7756b3419d077a7159d5fecbba3bd18c06ca0c6fa7aR14-R16). (`CONFIG_MODULES_DIFFERENTIAL_DRIVE` is actually outdated and doesn't exist anymore.)

Thanks @chfriedrich98 ! 🙏

sfuhrer reviewed (approved):
Good to go from my side. Please squash merge or clean up commits manually.

chfriedrich98 commented:
rebased and squashed commits

Pedro-Roque commented:
@chfriedrich98 nice contribution! Please consider setting up a SITL test for rover based on your own testing setup.
